a blog post about getting started with otel and rust (#1775)
open-telemetry/opentelemetry.io: content/en/blog/2022/rust-instrumentation/index.md

---

> title: Instrumenting a Rust application with OTel

Reviewer: ```suggestion
title: Instrumenting a Rust application
```

linkTitle: Rust Instrumentation
date: 2022-09-21
author: '[Yoav Danieli](https://github.com/ydan287) (Aspecto)'
canonical_url: https://www.aspecto.io/blog/distributed-tracing-with-opentelemetry-rust/
---

> # Getting Started: OpenTelemetry in Rust

Reviewer 2: ```suggestion
# Getting Started: OpenTelemetry in Rust
```
Seems this heading 1 is not very required since all legacy blogs have put this heading 1 into the front matter as a title.

Reviewer: Right, headings in all site page start at level 2, but in this case, the blog title is sufficient, just **drop this heading**.

Reviewer: @windsonsea - I'm not sure what your suggested change is, the two lines seem the same.

Reviewer 2: I mean the blogs normally have a style to take the title into front matter at top. Few blogs have a heading 1. 

You can check the blog folder to learn more.

![image](https://user-images.githubusercontent.com/79828097/201446892-863304f7-34e4-48e0-86be-00f0f12da4dd.png)

> ![Cover image showing Rust and OpenTelemetry Logos](cover-image.webp)

Reviewer: This is a very large image. Maybe make it a bit smaller?

In this article, I will share my experience adding OpenTelemetry distributed tracing to a Rust application.

I will attempt to answer the following questions:

- How to instrument a Rust application with OpenTelemetry?

- How to add instrumentations to a Rust application?

- How to use tracing to debug Rust applications?

- How to visualize traces and spans?

- How to preserve our span context in a multithreaded environment?

> - What are your must-know crates when it comes to tracing in Rust?

Reviewer: Nit: remove the blank lines between the bullet items.

To record spans and events we use macros, here are some examples:

```Rust
{
    // constructs a span with TRACE level.
    let span = span!(Level::TRACE, "my span");
    let _enter_guard = span.enter();

    // recording an INFO level event
    event!(Level::INFO, "inside the block");

    // the _enter_guard will be dropped here and the span will exit.
}
{
    let some_variable = 100_000;
    // the entered consumes the span. which means it now owns it instead of just borrowing it.
    let _enter_guard = info_span!("INFO level span").enter();

    // records an ERROR level event with some attributes
    error!(it = "accepts multiple arguments", and_also = ("multiple types"), how_many="?", could_be_more_than=some_variable);
}
```

Subscriber - For spans and events to be recorded and aggregated we need to initialize a Subscriber. In most cases, we will use an existing Subscriber implementation.

The crate [tracing_subscriber](https://docs.rs/tracing-subscriber/0.3.15/tracing_subscriber/) provides us with general use case subscriber that can be extended and modified according to specific needs.

We initialize the subscriber using Registry struct:

> ```shell

Reviewer: ```suggestion
```rust
```

let subscriber = tracing_subscriber::Registry::default()
```

This struct implements a subscriber and exposes us to another important feature of this crate - The Layer. Using layers, we can configure our subscriber to apply specific behaviors when interacting with spans and events.

For example, if we want to filter some of our data, export it, format it or edit it, we can create or use an existing layer and compose it with the registry subscriber.

> ```shell

Reviewer: ```suggestion
```rust
```

let subscriber = tracing_subscriber::Registry::default().with(SomeLayer)
```

We will see more examples on how to use these crates to instrument a Rust program in the practical section.

### OpenTelemetry in Rust

> The crate that provides support for OpenTelemetry is called - how convenient - [OpenTelemetry](https://docs.rs/opentelemetry/latest/opentelemetry/).

Reviewer: ```suggestion
The crate that provides support for OpenTelemetry is named [OpenTelemetry](https://docs.rs/opentelemetry/latest/opentelemetry/).
```

Let’s go over some of its key APIs:

> - Tracer. Inside the crate, there is a module called traces. It introduces us to a trait called Tracer. This is what tracks and connects our spans to create traces. To use Tracer, we create a subscriber layer. This is how the subscriber knows where to send the spans and how to generate traces.

Reviewer: ```suggestion
- **Tracer**. Inside the crate, there is a module called traces. It introduces us to a trait called Tracer. This is what tracks and connects our spans to create traces. To use Tracer, we create a subscriber layer. This is how the subscriber knows where to send the spans and how to generate traces.
```

> - Global. This module provides an API so that no matter in what section of the code we currently are, we can access the subscriber, tracer, and propagation context.

Reviewer: ```suggestion
- **Global**. This module provides an API so that no matter in what section of the code we currently are, we can access the subscriber, tracer, and propagation context.
```

> - SDK. This module provides an implementation for common use cases with OpenTelemetry. For example, it provides an exporter to send traces to stdout. It also implements context propagation methods that we can use instead of implementing ourselves.

Reviewer: ```suggestion
- **SDK**. This module provides an implementation for common use cases with OpenTelemetry. For example, it provides an exporter to send traces to stdout. It also implements context propagation methods that we can use instead of implementing ourselves.
```

> The git repository called [opentelemetry-rust](https://github.com/open-telemetry/opentelemetry-rust) contains implementations of several crates that expand the OpenTelemetry ecosystem. There you can find common instrumentations, exporters, and subscribers use them in a Rust program.

Reviewer: ```suggestion
The GitHub repository called [opentelemetry-rust](https://github.com/open-telemetry/opentelemetry-rust) contains implementations of several crates that expand the OpenTelemetry ecosystem. There you can find common instrumentations, exporters, and subscribers use them in a Rust program.
```

### Distributed tracing in Rust

Using crates from the [OpenTelemetry-Rust](https://github.com/open-telemetry/opentelemetry-rust) repo, we can instrument our application across multiple services and send them to a distributed platform.

> For example, we can use the [OpenTelemetry-jaeger](https://crates.io/crates/opentelemetry-jaeger) to send our traces to a remote Jaeger platform.

Reviewer 3: ```suggestion
For example, we can use the [opentelemetry-jaeger](https://crates.io/crates/opentelemetry-jaeger) to send our traces to a remote Jaeger platform.
```
I believe in cases where you're referring to a module the text doesn't need to be stylized. It follows the same logic as modules.

> Another example is the [opentelemetry-otlp](https://crates.io/crates/opentelemetry-otlp) crate.

Reviewer 4: please remove opentelemetry-jaeger and use opentelemetry-otlp, jaeger recommends switching to OTLP (cc [breedx-splk](https://github.com/breedx-splk))

// add the middleware to the server.
HttpServer::new(move || {
        App::new()
            .wrap(RequestTracing::new())
            .app_data(web::Data::new(pool.clone()))
            .service(get_users_service())
    })
    .bind(("127.0.0.1", 3000))?
    .run()
    .await
```

Now, when running the application and invoking the create user endpoint, we print spans to the terminal.

Trying to understand something from this print is quite hard. Besides, we are gathered here today to visualize traces, not to print them to the terminal.

So I changed the tracer to export traces to Jaeger via the OTLP exporter

### Visualizing traces with Jaeger

```bash
cargo add opentelemetry_jaeger --features "rt-tokio collector_client isahc_collector_client"
```

```Rust
let tracer = opentelemetry_jaeger::new_collector_pipeline()
        .with_endpoint("http://localhost:14268/api/traces")
        .with_isahc()
        .with_service_name(env::var("SERVICE_NAME").unwrap().to_string())
        .install_batch(opentelemetry::runtime::Tokio)
        .expect("Error - Failed to create tracer.");
let tracing_leyer = tracing_opentelemetry::layer().with_tracer(tracer);

> ```

Reviewer 4: same here, please use OTLP

When running the application once more, I can view my traces in Jaeger:

![Jaeger distributed tracing platform showing our Rust service trace](jaeger-platfrom.png)

Clicking the trace, I can view the spans and their data:

![Jaeger distributed tracing platform showing raw data of our Rust service trace](raw-data.png)

### Manually instrumenting Rust functions

Let’s add more instrumentations to our service.

> By using the macro #[instrument] on a function, we can create a span for this function.

Reviewer: ```suggestion
By using the macro `#[instrument]` on a function, we can create a span for this function.
```

We can do that by wrapping the web::block function with our own tracing function:

```Rust
fn traced_web_block<F, R>(f: F) -> impl Future<Output = Result<R, BlockingError>>
where
    F: FnOnce() -> R + Send + 'static,
    R: Send + 'static,
{
    let current_span = tracing::Span::current();
    web::block(move || current_span.in_scope(f))
}
```

Here we get our current span and invoking the closure within its scope.

Let’s view our traces now:

![View a trace in Jaeger](trace-view.png)

We can now get a full picture of the flow of our application.

## Conclusion

We reviewed how to add distributed tracing with OpenTelemetry to the Rust application. We explored how to add both manual instrumentation and library-provided instrumentations.

We used Jaeger to visualize the flow of our system and learned what to do when the flow is broken.

To sum up, I enjoyed working on this example. I think the approach Rust is taking regarding tracing is interesting and worth considering when writing a Rust program. Furthermore, the ecosystem is growing so I am sure more libraries will provide instrumentations soon.

> I hope you learned something new. Feel free to reach out with any questions you might have.

Reviewer: > Feel free to reach out ...

Let folks know how to reach out, e.g.:  Slack, GitHub discussions or issues. And provide link(s).

---
title: Instrumenting a rust application with OTel
linkTitle: Rust Instrumentation
date: 2022-09-21
author: '[Yoav Danieli](https://github.com/ydan287) (Aspecto)'
canonical_url: https://www.aspecto.io/blog/distributed-tracing-with-opentelemetry-rust/
---

# Getting Started: OpenTelemetry in Rust

![Cover image showing Rust and OpenTelemetry Logos](cover-image.webp)

In this article, I will share my experience adding OpenTelemetry distributed tracing to a Rust application.

I will attempt to answer the following questions:

> - How to instrument Opentelemetry in Rust?

Reviewer 4: ```suggestion
- How to instrument a rust application with OpenTelemetry?
```

> - How to add manual and auto instrumentations to a Rust application?

Reviewer 4: I don't see any auto instrumentation. I think it doesn't even exist (yet) for rust.
```suggestion
- How to add instrumentations to a Rust application?
```

- How to use tracing to debug Rust applications?

- How to visualize traces and spans?

- How to preserve our span context in a multithreaded environment?

- What are your must-know crates when it comes to tracing in Rust?

I will begin by saying that I love Rust as a programming language, and I've been experimenting with it for the past year. Although it's innovative and has a great ecosystem, it is not a secret that Rust has quite a steep learning curve. So even though it's an intro-level article, I will not dive deep into Rust features, syntax, or semantics unless it's somehow connected to our subject (Opentelemetry and distributed tracing).

## Opentelemetry and Distributed Tracing in Rust

### Tracing in Rust

> In Rust we have a great crate simply called “[tracing](https://docs.rs/tracing/latest/tracing/)”. It is our core framework for tracing a Rust program. The crates documentation outlines its core concepts and APIs which are composed of these terms - spans, events, and subscribers.

Reviewer 4: ```suggestion
In Rust we have a great crate simply called [tracing](https://docs.rs/tracing/latest/tracing/). It is our core framework for tracing a Rust program. The crates documentation outlines its core concepts and APIs which are composed of these terms - spans, events, and subscribers.
```

> Spans - In contrast to a log line, which represents a precise moment in time, a span denotes a period of time with a beginning and an end.
> 
> Events - An Event could be compared to the log records generated by unstructured logging code, but unlike a typical log line, Events may occur within spans.

Reviewer 4: ```suggestion
```

let's assume that your readers know what spans,events are.

To record spans and events we use macros, here are some examples:

```rust
{
    // constructs a span with TRACE level.
    let span = span!(Level::TRACE, "my span");
    let _enter_guard = span.enter();

    // recording an INFO level event
    event!(Level::INFO, "inside the block");

    // the _enter_guard will be dropped here and the span will exit.
}
{
    let some_veriable = 100_000;
    // the entered consumes the span. which means it now owns it instead of just borrowing it.
    let _enter_guard = info_span!("INFO level span").entered();

    // records an ERROR level event with some attributes
    error!(it = "accepts multiple arguments", and_also = ("multiple types"), how_many="?", could_be_more_than=some_veriable);
}
```

Subscriber - For spans and events to be recorded and aggregated we need to initialize a Subscriber. In most cases, we will use an existing Subscriber implementation.

> The crate “[tracing_subscriber](https://docs.rs/tracing-subscriber/0.3.15/tracing_subscriber/)” provides us with general use case subscriber that can be extended and modified according to specific needs.

Reviewer 4: ```suggestion
The crate [tracing_subscriber](https://docs.rs/tracing-subscriber/0.3.15/tracing_subscriber/) provides us with general use case subscriber that can be extended and modified according to specific needs.
```

We initialize the subscriber using Registry struct:

```shell
let subscriber = tracing_subscriber::Registry::default()
```

This struct implements a subscriber and exposes us to another important feature of this crate - The Layer. Using layers, we can configure our subscriber to apply specific behaviors when interacting with spans and events.

For example, if we want to filter some of our data, export it, format it or edit it, we can create or use an existing layer and compose it with the registry subscriber.

```shell
let subscriber = tracing_subscriber::Registry::default().with(SomeLayer)
```

We will see more examples on how to use these crates to instrument a Rust program in the practical section.

### OpenTelemetry in Rust

> The crate that provides support for OpenTelemetry is called - how convenient - “[OpenTelemetry](https://docs.rs/opentelemetry/latest/opentelemetry/)”.

Reviewer 4: ```suggestion
The crate that provides support for OpenTelemetry is called - how convenient - [OpenTelemetry](https://docs.rs/opentelemetry/latest/opentelemetry/).
```

> The crate that provides support for OpenTelemetry is called - how convenient - “[OpenTelemetry](https://docs.rs/opentelemetry/latest/opentelemetry/)”.

Reviewer 3: ```suggestion
The crate that provides support for OpenTelemetry is called - how convenient - [opentelemetry](https://docs.rs/opentelemetry/latest/opentelemetry/).
```
Oddly enough it is the convenience crate 😅 we do have opentelemetry-api and opentelemetry-sdk as well.

> “OpenTelemetry provides a single set of APIs, libraries, agents, and collector services to capture distributed traces and metrics from your application.”

Reviewer 4: ```suggestion
```

Let’s go over some of its key APIs:

> - Tracer. Inside the crate, there is a module called traces. It introduces us to a trait called Tracer. This is what tracks and connects our spans to create traces. To use Tracer, we create a subscriber layer. This is how the subscriber knows where to send the spans and how to generate traces.

Reviewer 3: @svrnm  For modules other code references, should the form `traces` be used?

Reviewer 4: yes, makes it easier to see

Reviewer 5: > there is a module called traces

This use of the word "traces" is different from this use of the word:

> how to generate traces.

And so if you make it visually distinct either via a link, or markdown code formatting, etc. it will be more clear.

Reviewer 3: @NavehMevorach 
The way to fix this would be that when you're referring to the model encapsulate it
```suggestion
- Tracer. Inside the crate, there is a module called `traces`. It introduces us to a trait called Tracer. This is what tracks and connects our spans to create traces. To use Tracer, we create a subscriber layer. This is how the subscriber knows where to send the spans and how to generate traces.
```

- Global. This module provides an API so that no matter in what section of the code we currently are, we can access the subscriber, tracer, and propagation context.

> - SDK. This module provides an implementation for common use cases with OpenTelemetry. For example, it provides an exporter to send traces to stdout. It also implements context propagation methods that we can use instead of implementing ourselves.

Reviewer 6: Maybe worth calling out that SDK is in its separate crate since opentelemery 0.18

Reviewer 3: 💯 https://github.com/open-telemetry/opentelemetry.io/pull/1775#discussion_r997699808

The git repository called [opentelemetry-rust](https://github.com/open-telemetry/opentelemetry-rust) contains implementations of several crates that expand the opentelemetry ecosystem. There you can find common instrumentations, exporters, and subscribers use them in a Rust program.

### Distributed tracing in Rust

Using crates from the [opentelemetry-rust](https://github.com/open-telemetry/opentelemetry-rust) repo, we can instrument our application across multiple services and send them to a distributed platform.

For example, we can use the [opentelemetry-jaeger](https://crates.io/crates/opentelemetry-jaeger) to send our traces to a remote Jaeger platform.

Another example is the [OpenTelementy-otlp](https://crates.io/crates/opentelemetry-otlp) crate.

Traces in OTLP format can be exported to the OpenTelemetry Collector using the exporter in this crate. OpenTelemetry Collector accepts, processes, and exports traces in a vendor-agnostic manner.

## Practical Example

> For this example, I wanted to build a simple system I could use to integrate and visualize traces. I found auto-instrumentations for [Actix-Web](https://actix.rs/) Framework and [Diesel](https://diesel.rs/), an ORM and Query builder utility for SQL-based databases.

Reviewer 3: "instremented libraries"

> I chose to use these tools and build a simple REST api service to create, read and delete users following the examples given by these frameworks.

Reviewer 3: ```suggestion
I chose to use these tools and build a simple REST API service to create, read and delete users following the examples given by these frameworks.
```

Here is a link for the [source code](https://github.com/aspecto-io/opentelemetry-examples/tree/master/rust).

### Creating the service

> First, I followed the Actix web and diesel documentation and created a simple web server that exposes a user service and communicates with a local Postgres database:

Reviewer 3: ```suggestion
First, I followed the Actix-Web and Diesel documentation and created a simple web server that exposes a user service and communicates with a local Postgres database:
```

    Ok(HttpResponse::Ok().json(user))
}

pub fn get_users_service() -> actix_web::Scope {
    web::scope("/users")
        .service(get_users)
        .service(get_user_by_id)
        .service(create_user)
        .service(delete_user)
}

```

### Create our tracing setup

Now that I have a working service, I can add telemetry data. I will need to configure a tracing subscriber and its layers.

```rust
use opentelemetry::sdk::export::trace::stdout;
use opentelemetry::sdk::propagation::TraceContextPropagator;
use opentelemetry::sdk::{ trace, Resource };
use opentelemetry::KeyValue;
use opentelemetry::{ global };
use std::collections::HashMap;
use std::env;
use tracing_subscriber::{ prelude::*, EnvFilter, Registry};

pub fn init_telemetry() {

    // Define stdout tracer
    let stdout_tracer = stdout::new_pipeline().install_simple();

    // Define subscriber with a tracing layer to use our tracer
    let subscriber = Registry::default();

    // Layer to filter traces based on level - trace, debug, info, warn, error.
    let env_filter = EnvFilter::try_from_default_env().unwrap_or(EnvFilter::new("INFO"));
    // Layer to add our configured tracer.

>     let tracing_leyer = tracing_opentelemetry::layer().with_tracer(tracer);

Reviewer 3: ```suggestion
    let tracing_layer = tracing_opentelemetry::layer().with_tracer(tracer);
```
Nit

        .init();
}
```

This configuration adds an env_filter layer and a tracing layer. It prints to stdout all spans that their level is equal or above “info” log level.

### Adding Instrumentation for Actix-Web framework

Now let's add some spans by instrumenting the actions we take. Let's start by instrumenting the operation taken by Actix web.

For that, I will use the actix_web_opentelemetry crate which provides a middleware called RequestTracing. As its name suggests, this middleware extracts trace information from the request and creates a span with the same context.

```rust
use actix_web_opentelemetry::RequestTracing;
// add the middleware to the server.
HttpServer::new(move || {
        App::new()
            .wrap(RequestTracing::new())
            .app_data(web::Data::new(pool.clone()))
            .service(get_users_service())
    })
    .bind(("127.0.0.1", 3000))?
    .run()
    .await
```

Now, when running the application and invoking the create user endpoint, we print spans to the terminal.

Trying to understand something from this print is quite hard. Besides, we are gathered here today to visualize traces, not to print them to the terminal.

> So I changed the tracer to export traces to Jaeger..

Reviewer 4: ```suggestion
So I changed the tracer to export traces to Jaeger via the OTLP exporter
```

Please use the OTLP exporter.

### Visualizing traces with Jaeger

```bash
cargo add opentelemetry_jaeger --features "rt-tokio collector_client isahc_collector_client"
```

```rust
let tracer = opentelemetry_jaeger::new_collector_pipeline()
        .with_endpoint("http://localhost:14268/api/traces")
        .with_isahc()
        .with_service_name(env::var("SERVICE_NAME").unwrap().to_string())
        .install_batch(opentelemetry::runtime::Tokio)
        .expect("Error - Failed to create tracer.");

> let tracing_leyer = tracing_opentelemetry::layer().with_tracer(tracer);

Reviewer 4: please use the OTLP exporter.
